pkg/metrics/metrics.go

```diff
@@ -0,0 +1,136 @@
+// Package metrics provides Prometheus metrics for CoreDNS probes.
+package metrics
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// ProbeMetrics holds all metrics for CoreDNS probe.
+type ProbeMetrics struct {
+	// Metrics for DNS queries per CoreDNS endpoint
+	totalQueriesPerEndpoint  *prometheus.CounterVec
+	failedQueriesPerEndpoint *prometheus.CounterVec
+	rttHistogram             *prometheus.HistogramVec
+
+	// Registry for all metrics
+	registry *prometheus.Registry
+
+	// HTTP server for exposing metrics
+	server *http.Server
+
+	// Lock for managing server state
+	mu      sync.Mutex
+	running bool
+}
+
+// New creates a new ProbeMetrics instance with registered Prometheus metrics.
+func New() *ProbeMetrics {
+	registry := prometheus.NewRegistry()
+
+	totalQueries := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "coredns_probe_queries_total",
+			Help: "Total number of DNS queries sent to CoreDNS endpoints",
+		},
+		[]string{"endpoint"},
+	)
+
+	failedQueries := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "coredns_probe_queries_failed_total",
+			Help: "Number of failed DNS queries to CoreDNS endpoints",
+		},
+		[]string{"endpoint"},
+	)
+
+	rttHistogram := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "coredns_probe_rtt_milliseconds",
+			Help:    "Histogram of round-trip time for successful DNS queries in milliseconds",
+			Buckets: []float64{0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5, 10, 20, 50, 100, 200, 500, 1000},
```

Comment:
histograms already do counts so theoretically you just need this with a succes label and you can drop other two.